[AIRFLOW-224] Collect orphaned tasks and reschedule them

bolkedebruin opened the pull request:
Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- https://issues.apache.org/jira/browse/AIRFLOW-224

Tasks can get orphaned if the scheduler is killed in the middle
of processing the tasks or if the MQ queue is cleared without
a worker having picked these up. Now tasks do not get set
to a scheduled state anymore if they have not been sent to the
executor yet. Next to that a garbage collector scans the executor
for tasks not being present and reschedules those if needed.

plypaul commented on airflow/jobs.py:
```diff
@@ -668,7 +657,28 @@ def _do_dags(self, dagbag, dags, tis_out):
                 self.logger.exception(e)
 
+    @provide_session
+    def _garbage_collect(self, dag_run, session=None):
```
Since garbage collection is often referred in the memory context, can we rename this? Maybe `_reset_state_for_orphaned_tasks`?

bolkedebruin committed:
[AIRFLOW-224] Collect orphaned tasks and reschedule them

Tasks can get orphaned if the scheduler is killed in the middle
of processing the tasks or if the MQ queue is cleared without
a worker having picked these up. Now tasks do not get set
to a scheduled state anymore if they have not been sent to the
executor yet. Next to that a garbage collector scans the executor
for tasks not being present and reschedules those if needed.

plypaul commented on airflow/jobs.py:
```diff
@@ -738,8 +749,20 @@ def _execute(self):
+                        ti.refresh_from_db(session=session, lock_for_update=True)
+                        if ti.state == State.SCHEDULED:
+                            session.commit()
+                            self.logger.debug("Task {} was picked up by another scheduler"
```
I don't think we support multiple schedulers officially? Without locks, there a lot of race conditions.

bolkedebruin replied:
There are multiple people that do run multiple schedulers and we don't officially discourage it either. The updates to the scheduler I made make it almost safe to do so (only UP_FOR_RETRY tasks remain to have that issue).

plypaul commented on airflow/jobs.py:
```diff
@@ -687,6 +695,16 @@ def _execute(self):
+            session=session
+        )
+        for dr in active_runs:
+            self._reset_state_for_orphaned_tasks(dr, session=session)
```
This turns out to be quite slow - we have several thousand DAG runs, so when it starts, several minutes are spent going through this loop.

criccomini replied:
@plypaul curious--did you guys work around this? Is there a patch to fix?

bolkedebruin replied:
It shouldnt break a sweat on a couple of thousand. Where exactly is it slow? In the db or in Python?

bolkedebruin replied:
I'll check if it is proper indexed in the db. If it isn't it might need one. Otherwise I can rework it to fully run in the db in one go. That can happen in one statement I think. If you @plypaul can let me know some metrics that would make it easier

criccomini replied:
@bolkedebruin pretty sure this is related to the 'add index on task state' PR that's in flight right now :)

bolkedebruin replied:
@criccomini @plypaul think so too, I double checked and I think the query that is being run indeed will not use an index (index is on: dag_id, task_id, execution_date, state - task_id is missing for this query so it wont hit).
